### ISSUE TITLE
Handle 3rd party NPM dependencies in source-build

### ISSUE THREAD

**Issue by wtgodbe** [dotnet/aspnetcore#53710]
We recently attempted to turn on the build of the NPM components in source-build. This led to issues as the NPM/node components have a couple 3rd party dependencies, which can't be restored in offline source-build. We have a couple options for how to tackle this:

1. Check in a node package cache (matching our package.lock.json file) to a separate repo, which is a submodule of aspnetcore. Any time a dev changes a node dependency, they should re-generate the cache, push it to the new repo, and update the submodule in their PR. We'll need  to exclude that repo from automated submodule updates in that case (dependabot). Node has a validation mode that we can run during the build to ensure the package cache matches what's in package.lock.json.

2. Add a pre-step for source build that does an online restore before we start offline source-build. There are already online pre-steps for source-build, so this wouldn't be against the "spirit" of source-build.

We should choose whichever approach most closely matches what our source-build partners do.

CC @dotnet/distro-maintainers @javiercn @mthalman @MichaelSimons

**Comment by wtgodbe** [comment]
> Regarding these dependencies, one thing you'll need to be aware of is the licenses of these packages. I just saw that we have a license scan failure for the VMR coming from aspnetcore's package-lock.json file. A new jackspeak dependency was added which has a license called Blue Oak ([Blue Oak Model License 1.0.0 (blueoakcouncil.org)](https://blueoakcouncil.org/license/1.0.0)). Our distro maintainer partners require that all referenced licenses be free and open source. It's not immediately clear whether this license fits that requirement. Related issue: https://github.com/dotnet/source-build/issues/4014

**Comment by omajid** [comment]
IMO, the licensing of the dependencies should be a separate issue. We could take a dependency on Node.js, or check in pre-compiled JavaScript code. We could have a package cache as a sub-repo or require maintainers to run a prep.sh-like thing themselves. But none of it would change the fact that we (distro maintainers) would be pulling in something with a potential licensing issue.
